src/overlays/actors/ovl_Bg_Haka_Ship/z_bg_haka_ship.c

```diff
@@ -41,25 +39,203 @@ const ActorInit Bg_Haka_Ship_InitVars = {
+void BgHakaShip_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgHakaShip* this = THIS;
+    s32 pad;
+    CollisionHeader* colHeader;
+
+    colHeader = NULL;
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+    DynaPolyActor_Init(&this->dyna, 1);
+    this->switchFlag = (thisx->params >> 8) & 0xFF;
+    this->dyna.actor.params &= 0xFF;
+
+    if (this->dyna.actor.params == 0) {
+        CollisionHeader_GetVirtual(&D_0600E408, &colHeader);
+        this->counter = 8;
+        this->actionFunc = BgHakaShip_WaitForSong;
+    } else {
+        CollisionHeader_GetVirtual(&D_0600ED7C, &colHeader);
+        this->actionFunc = BgHakaShip_ChildUpdatePosition;
+    }
+    this->dyna.bgId = DynaPoly_SetBgActor(globalCtx, &globalCtx->colCtx.dyna, &this->dyna.actor, colHeader);
+    this->dyna.actor.world.rot.y = this->dyna.actor.shape.rot.y - 0x4000;
+    this->yOffset = 0;
+    if (this->dyna.actor.params == 0 &&
+        !Actor_SpawnAsChild(&globalCtx->actorCtx, &this->dyna.actor, globalCtx, ACTOR_BG_HAKA_SHIP,
+                            this->dyna.actor.world.pos.x + -10.0f, this->dyna.actor.world.pos.y + 82.0f,
+                            this->dyna.actor.world.pos.z, 0, 0, 0, 1)) {
+        Actor_Kill(&this->dyna.actor);
+    }
+}
+
+void BgHakaShip_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgHakaShip* this = THIS;
+
+    DynaPoly_DeleteBgActor(globalCtx, &globalCtx->colCtx.dyna, this->dyna.bgId);
+    func_800F89E8(&this->bellSoundPos);
+}
+
+void BgHakaShip_ChildUpdatePosition(BgHakaShip* this, GlobalContext* globalCtx) {
+    BgHakaShip* parent;
+
+    parent = (BgHakaShip*)this->dyna.actor.parent;
+    if (parent && parent->dyna.actor.update) {
+        this->dyna.actor.world.pos.x = parent->dyna.actor.world.pos.x + -10.0f;
+        this->dyna.actor.world.pos.y = parent->dyna.actor.world.pos.y + 82.0f;
+        this->dyna.actor.world.pos.z = parent->dyna.actor.world.pos.z;
+    } else {
+        this->dyna.actor.parent = NULL;
+    }
+}
+
+void BgHakaShip_WaitForSong(BgHakaShip* this, GlobalContext* globalCtx) {
+    if (Flags_GetSwitch(globalCtx, this->switchFlag)) {
+        if (this->counter) {
+            this->counter--;
+        }
+        if (this->counter == 0) {
+            this->counter = 130;
+            this->actionFunc = BgHakaShip_CutsceneStationary;
+            osSyncPrintf("シーン 外輪船 ...  アァクション！！\n");
+            func_800800F8(globalCtx, 3390, 999, &this->dyna.actor, 0);
+        }
+    }
+}
+
+void BgHakaShip_CutsceneStationary(BgHakaShip* this, GlobalContext* globalCtx) {
+    if (this->counter) {
+        this->counter--;
+    }
+    this->yOffset = sinf(this->counter * (M_PI / 25)) * 6144.0f;
+    if (this->counter == 0) {
+        this->counter = 50;
+        this->actionFunc = BgHakaShip_Moving;
+    }
+}
+
+void BgHakaShip_Moving(BgHakaShip* this, GlobalContext* globalCtx) {
+    f32 distanceFromHome;
+    BgHakaShip* child;
```

Comment:
This could probably just be `Actor*` type but I'd wait for others to give input.

Reply:
I agree so long as it doesn't need to access any extra variables beyond the base actor